### PR TITLE
Big endian support

### PR DESCRIPTION
**Pull request by laanwj**
Fix issue #888.

This has been structured so that each compatibility change is one commit that touches only one file. After the initial build change, they are independent.

Most extensive changes are in **'src/serialize.h: base serialization level endianness neutrality'**. I had to replace READDATA and WRITEDATA with functions that take sized integer types to make use of the proper `endian.h` functions. I'm confident that the end result is the same, although this may require more tests.

I've tested this on mipsbe32. 
- All tests pass
- Testnet syncs correctly
- Node can successfully function on P2P mainnet
- Checked that data directory can be copied between endians with no adverse results (only peers.dat required special attention here)

Known issues (to be fixed before merge):
- ~~DNS seeding always comes with 0 results on BE~~ (confirmed as working by @paveljanik on real hardware, must have been issue with my qemu-user setup)

### CODE REVIEW

**Comment by paveljanik** [comment]
Mac Mini PPC with 512MB RAM just finished testnet sync. Debian GNU/Linux 7.7. No issues so far. Even with few manual stops during the IBD. Generating new wallet was extremely slow... Few compile warnings for signed/unsigned and always true (I will investigate them later).

File: src/arith_uint256.cpp [hunk 1]
```diff
@@ -10,6 +10,7 @@
 
 #include <stdio.h>
 #include <string.h>
+#include "crypto/common.h"
```
Comment (Diapolo) [inline src/arith_uint256.cpp]: I just want to know if we completly stop doing any include ordering/cleanup or even stop following the style we currently have in most files? If the new philosophy is new features before everything else I'm losing the last fun for this project :-/.

Comment (sipa) [inline src/arith_uint256.cpp]: I'm sorry to hear that you don't find fun in many of the much more interesting things you've been doing before, but no - we removed this from the policy because the cost is just not worth the benefits. It requires too much time from maintainers and people writing patches for a very tiny (but non-zero) benefit, so we reverted it for that reason.

Comment (jgarzik) [inline src/arith_uint256.cpp]: @Diapolo Every code cleanup -- including those authored by myself -- has review costs and potentially disrupts more important functional work that actually changes behavior for the users.

Style takes a back seat to more practical changes, as it must.  It sucks when it's your changes.

Include file ordering is just not important enough to us to care about maintaining it in our own changes.  We absolutely value your contributions, but you have to understand how to work with everyone else in the project and judge your impact on other developers, not just the codebase.

We don't have rules, we have guidelines.

**Comment by theuni** [comment]
utACK, nothing to complain about.

Do you have an idea where the DNS issue lies?

**Comment by theuni** [comment]
Mmm.. on second thought, I suppose I'd first like to know the result of moving a fully-sync'd .bitcoin dir from a BE machine to a LE one, or the other way around.

**Comment by laanwj** [comment]
@theuni As I mention in the OP, I've tested that, and it worked. UTXO databases and block databases are compatible between BE and LE.

I haven't had time to look into the DNS seeding issue.

**Comment by paveljanik** [comment]
DNS seeding works OK here on PPC Mac Mini:
```
2015-01-08 07:42:38 init message: Loading addresses...
2015-01-08 07:42:38 ERROR: Read : Failed to open file /home/bitcoin/.bitcoin/testnet3/peers.dat
2015-01-08 07:42:38 Invalid or missing peers.dat; recreating
2015-01-08 07:42:38 Loaded 0 addresses from peers.dat  43ms
...
2015-01-08 07:42:38 Loading addresses from DNS seeds (could take a while)
2015-01-08 07:42:38 ERROR: AcceptToMemoryPool : inputs already spent
2015-01-08 07:42:38 ERROR: AcceptToMemoryPool : inputs already spent
2015-01-08 07:42:48 Added 27 addresses from 95.85.12.7: 0 tried, 27 new
2015-01-08 07:42:48 Added 6 addresses from ::: 0 tried, 33 new
2015-01-08 07:42:48 Added 1 addresses from ::: 0 tried, 34 new
2015-01-08 07:42:49 Added 28 addresses from ::: 0 tried, 62 new
2015-01-08 07:42:49 65 addresses found from DNS seeds
2015-01-08 07:42:49 dnsseed thread exit
```

**Comment by paveljanik** [comment]
I'm moving the bitcoin directory there and back between LE and BE host without issues so far.

File: src/arith_uint256.cpp (hunk 1 above)
Comment (jonasschnelli) [inline src/arith_uint256.cpp]: Maybe this is the wrong place for this discussion. But better here then nowhere.
IMO clean code leads to good software quality in general.
Once the code gets a mess, the quality in general will follow.
"A broken window is the start of a vagabonded house". Thats why i honor @Diapolo work on one hand.

But:
1.)
Nowadays, code-structure can be handled by tools like clang-format. There is no need for nits for spaces, indents, etc.
Even if the clang-format output is not stable, i think a re-formatting from time-to-time over the whole code makes sense and will prevent us from thousands of nits and gives us more time to focus on the "real stuff".

2.)
Alphabetic ordering of includes just don't make sense. Someone does it, someone not. I think we don't need nit's for this.

3.)
Cleanup commits like #5616 are okay, but will break (requires rebase) all other and upcoming pull requests in this area. That's why i would recommend to keep these kind of pull request to a minimum. Better provide a large cleanup commit once (at best in a code-quiet time).
To much of cleanup-commit will slow down important work.

**Comment by theuni** [comment]
@laanwj Ah, I missed that part in the description. Thanks for verifying.

@paveljanik Thanks to you as well.

File: src/arith_uint256.cpp (hunk 1 above)
Comment (zander) [inline src/arith_uint256.cpp]: Include ordering is there for a very good purpose, which is to ensure stand-alone headers. This purpose is just not really all that relevant in a project whose headers are not installed and exported for 3rd party use.  Like a DLL based project would need it.

tl;dr dropping the requirement for header-inclusion-ordering makes sense to me.

**Comment by laanwj** [comment]
Had to rebase over the openssl fix (no code changes or conflicts) to make tests work again.

**Comment by laanwj** [comment]
Added the missing float/double serialization tests and removed the WIP tag.

**Comment by ry60003333** [comment]
Everything seems to be working here as well! I am using a PowerMac G5 with Debian 7; I've been waiting for big endian support for a long time, and can't wait until this is merged into the main release!

**Comment by sipa** [comment]
Untested ACK.

We do have a few special-cased pieces of code for little-endian vs generic. Do we need all of them? I would suggest to either:
* get rid of all WORDS_BIGENDIAN ifdefs (except in compat), and use the generic code everywhere
* have a mechanism to enable generic code at compile-time, and do at least one automated test with generic code.

**Comment by ry60003333** [comment]
If any core developers would like, I can provide full root SSH access to a Debian 7 PowerPC system for testing purposes to make sure future builds work on big endian systems. I would love for this to be integrated in a 0.10 release.

**Comment by sipa** [comment]
@laanwj I would be fine with just getting rid of all LE optimized code versions here, and later benchmark whether any are worth it to add back.

**Comment by laanwj** [comment]
@ry60003333 This is too late for the 0.10 release, but can make it to 0.11 in july

@sipa OK, fine with me, there are two cases of WORDS_BIGENDIAN left and neither of them seems to be in a performance critical place.

**Comment by sipa** [comment]
Untested ACK.

**Comment by ry60003333** [comment]
@laanwj 0.11 would work for me! I've been looking forward to being able to run full nodes on IBM POWER servers for a long time now!

File: src/script/script.h [hunk 2]
```diff
@@ -14,6 +14,7 @@
 #include <string.h>
 #include <string>
 #include <vector>
+#include "crypto/common.h"
```
Comment (Diapolo) [inline src/script/script.h]: So we are now also don't try to differentiate the blocks by own headers, standard headers and library headers?
